### PR TITLE
Embedded webextension isolated

### PR DESCRIPTION
**Pull request by Hainish**

### CODE REVIEW

**Comment by Hainish** [comment]
A failure to the `firefox` tests is expected, resolution depends on https://github.com/EFForg/https-everywhere/pull/11789.  This is rebased on top of https://github.com/EFForg/https-everywhere/pull/11791

File: chromium/background.js [hunk 1]
```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+/**
+ * Import extension settings (custom rulesets, ruleset toggles, globals) from an object
+ * @param settings the settings object
+ */
+async function import_settings(settings){
+  if(settings.changed){
+    // Load custom rulesets and add to storage
+    await new Promise(resolve => {
+      for(let ruleset of settings.custom_rulesets){
+        all_rules.addFromXml((new DOMParser()).parseFromString(ruleset, 'text/xml'));
```
Comment (jsha) [inline chromium/background.js]: Please factor out a function load_legacy_custom_rulesets, so that you can call it here after you do `storage.set`.

File: chromium/background.js [hunk 2]
```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+          resolve(false);
+        })
+      })(ruleset_name, settings.rule_toggle[ruleset_name]));
+    }
+    await Promise.all(rule_toggle_promises);
+
+    all_rules.ruleCache = new Map();
+
+    // Set/store globals
+    storage.set({'httpNowhere': settings.prefs.http_nowhere_enabled});
```
Comment (jsha) [inline chromium/background.js]: Is there a reason you provide a callback when calling set above, but not here? It seems like we should use the callback on both, to detect success or failure.

File: chromium/background.js [hunk 3]
```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+              }
+            }
+          }
+          resolve(false);
+        })
+      })(ruleset_name, settings.rule_toggle[ruleset_name]));
+    }
+    await Promise.all(rule_toggle_promises);
+
+    all_rules.ruleCache = new Map();
```
Comment (jsha) [inline chromium/background.js]: Can you add a comment explaining why this is needed?

File: chromium/background.js [hunk 4]
```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+    });
+
+    // Load all the ruleset toggles into memory and store
+    let rule_toggle_promises = [];
+    for(let ruleset_name in settings.rule_toggle){
+      localStorage[ruleset_name] = settings.rule_toggle[ruleset_name];
+
+      rule_toggle_promises.push((ruleset_name, ruleset_active => {
+        return new Promise(resolve => {
+          for(let host in all_rules.targets){
```
Comment (jsha) [inline chromium/background.js]: A couple of comments here:

 - This doesn't seem to do any async work, so why is it wrapped in a Promise?
 - This seems to re-do the work of marking rulesets active or inactive that should happen at startup. Similar to the loading of legacy rulesets I mentioned above, can you just store the settings, and then re-load the rulesets? In fact, since loading the rulesets should include loading legacy_custom_rulesets, you can just do that and don't have to separately invoke the loader for legacy_custom_rulesets.

File: chromium/background.js [hunk 5]
```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+      })(ruleset_name, settings.rule_toggle[ruleset_name]));
+    }
+    await Promise.all(rule_toggle_promises);
+
+    all_rules.ruleCache = new Map();
+
+    // Set/store globals
+    storage.set({'httpNowhere': settings.prefs.http_nowhere_enabled});
+    storage.set({'showCounter': settings.prefs.show_counter});
+    isExtensionEnabled = settings.prefs.global_enabled;
```
Comment (jsha) [inline chromium/background.js]: This setting doesn't seem to get stored?

File: makexpi.sh [hunk 6]
```diff
@@ -101,7 +100,25 @@ fi
 # Prepare packages suitable for uploading to EFF and AMO, respectively.
 [ -d pkg ] || mkdir pkg
 rsync -aL --delete --delete-excluded --exclude /chrome/content/rules src/ pkg/xpi-eff
-cp -a translations/* pkg/xpi-eff/chrome/locale/
+
+# The next lines are for embedded WebExtensions, and can be deleted after the full transition to WebExtensions
```
Comment (jsha) [inline makexpi.sh]: You should probably include a footer indicating the end of the lines that can be deleted after the full transition.

File: src/bootstrap.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+function startup({webExtension}) {
+  webExtension.startup().then(api => {
+    const {browser} = api;
+    browser.runtime.onMessage.addListener((msg, sender, sendReply) => {
+      if (msg == "import-legacy-data") {
+        let globals = get_prefs("extensions.https_everywhere.");
+        let prefs_array = globals.getChildList("", {});
+        if(~prefs_array.indexOf("webextensions-migrated") ? !globals.getBoolPref("webextensions-migrated") : true){
```
Comment (jsha) [inline src/bootstrap.js]: This if statement is overly complicated. In particular, `:?` operators are ideally avoided because they are hard to read and error prone. Embedding a `:?` inside an if makes that especially so. :-) Can you break it out for clarity? Also, the ~ operator is a new idiom in this codebase. I think it would be better not to introduce it in such a large change without strong motivation.

File: src/bootstrap.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+          let rules = get_prefs("extensions.https_everywhere.rule_toggle.");
+
+          let changed = false;
+
+          let prefs = {
+            http_nowhere_enabled:  ~prefs_array.indexOf("http_nowhere.enabled") ? globals.getBoolPref("http_nowhere.enabled") : false,
```
Comment (jsha) [inline src/bootstrap.js]: This pattern of "check if pref exists, return default value if not" should be a small helper function.

File: src/bootstrap.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+function startup({webExtension}) {
+  webExtension.startup().then(api => {
+    const {browser} = api;
+    browser.runtime.onMessage.addListener((msg, sender, sendReply) => {
+      if (msg == "import-legacy-data") {
+        let globals = get_prefs("extensions.https_everywhere.");
+        let prefs_array = globals.getChildList("", {});
+        if(~prefs_array.indexOf("webextensions-migrated") ? !globals.getBoolPref("webextensions-migrated") : true){
+          let rules = get_prefs("extensions.https_everywhere.rule_toggle.");
```
Comment (jsha) [inline src/bootstrap.js]: Move `let rules` and `let changed` down to right before they are used.

**Comment by koops76** [comment]
@Hainish Also, can you already code this in StandardJS style or the style of your choice (I recommend Standard)?

File: chromium/background.js [hunk 10]
```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
     updateState();
   } else if (message.type == "remove_rule") {
     removeRule(message.object);
+  } else if (message.type == "import_settings") {
```
Comment (jsha) [inline chromium/background.js]: Can you add a comment explaining when this is used? I.e., importing from the Options page rather than automatically from the parent extension?

File: chromium/background.js (hunk 4 above)
Comment (Hainish) [inline chromium/background.js]: This task can take quite a while, since it's iterating through all ruleset targets and checking all rules they contain until finding a match.  Creating a new promise for each of these isn't to resolve callbacks, but to start asynchronous program flow for each of these potentially many toggles.

Comment (Hainish) [inline chromium/background.js]: Since this is javascript and not actually dedicating separate cores to the problem, it does not necessarily give us any efficiency gain to split this up into multiple iterations of the event loop.  I can refactor this to be synchronous.

File: chromium/background.js (hunk 5 above)
Comment (Hainish) [inline chromium/background.js]: It's true, the extension enabled toggle is not stored (and has never been).  We should fix that but I figured it was outside of the scope of this PR.

File: src/bootstrap.js (hunk 9 above)
Comment (Hainish) [inline src/bootstrap.js]: I agree for `rules`, but for `changed` it seems this is a var which is featured prominently throughout this scope.  I'd argue that it's appropriate to keep it at the top.

File: chromium/background.js (hunk 5 above)
Comment (jsha) [inline chromium/background.js]: Oh, interesting! I did not know that. It seems weird to apply this toggle only on first run after an upgrade, though. If storing it doesn't work, I'd suggest just removing this line.

File: src/bootstrap.js (hunk 9 above)
Comment (jsha) [inline src/bootstrap.js]: Sounds reasonable.

File: chromium/background.js (hunk 10 above)
Comment (Hainish) [inline chromium/background.js]: Added.

File: chromium/background.js [hunk 11]
```diff
@@ -674,5 +682,56 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+      storage.set({"legacy_custom_rulesets": settings.custom_rulesets}, resolve);
+    });
+    load_legacy_custom_rulesets(settings.custom_rulesets);
+
+    // Load all the ruleset toggles into memory and store
+    let rule_toggle_promises = [];
+    for(let ruleset_name in settings.rule_toggle){
+      localStorage[ruleset_name] = settings.rule_toggle[ruleset_name];
+
+      (ruleset_name, ruleset_active => {
```
Comment (jsha) [inline chromium/background.js]: Did you see my earlier comment about just storing the data and reloading? E.g. `all_rules = new RuleSets(localStorage)`

File: chromium/background.js [hunk 12]
```diff
@@ -674,5 +682,56 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+            if(ruleset.name == ruleset_name){
+              ruleset.active = ruleset_active;
+              return;
+            }
+          }
+        }
+      })(ruleset_name, settings.rule_toggle[ruleset_name]);
+    }
+
+    // This is necessary in order to clear the cache for any previous ruleset lookups
```
Comment (jsha) [inline chromium/background.js]: If you do the reload, this won't be necessary anymore.

File: chromium/background.js [hunk 13]
```diff
@@ -674,5 +682,56 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+    all_rules.ruleCache = new Map();
+
+    // Set/store globals
+    await new Promise(resolve => {
+      storage.set({'httpNowhere': settings.prefs.http_nowhere_enabled}, resolve);
+    });
+    await new Promise(resolve => {
+      storage.set({'showCounter': settings.prefs.show_counter}, resolve);
+    });
+    isExtensionEnabled = settings.prefs.global_enabled;
```
Comment (jsha) [inline chromium/background.js]: Did you see my comment about removing this since it will only work once, and is therefore inconsistent/confusing?

File: chromium/background.js (hunk 11 above)
Comment (Hainish) [inline chromium/background.js]: Ah! Didn't catch that.  Fixing now.

**Comment by Hainish** [comment]
Ready for review again.

**Comment by Hainish** [comment]
Hold on, for some reason the default rulesets are absent when the second call of `new RuleSets(localStorage)` is made.

**Comment by koops76** [comment]
Do you mutate the argument in the `RuleSets` constructor?

**Comment by Hainish** [comment]
@koops76 not in a way that would cause targets to be lost

**Comment by koops76** [comment]
Try making a copy of everything in `localStorage` and passing that copy into the `RuleSets` constructor.

File: chromium/background.js [hunk 14]
```diff
@@ -674,5 +674,44 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+
+    // Load all the ruleset toggles into memory and store
+    let rule_toggle_promises = [];
+    for(let ruleset_name in settings.rule_toggle){
+      localStorage[ruleset_name] = settings.rule_toggle[ruleset_name];
+    }
+    all_rules = new RuleSets(localStorage);
+
+    // Set/store globals
+    await new Promise(resolve => {
```
Comment (koops76) [inline chromium/background.js]: DRY

File: chromium/options.css [hunk 15]
```diff
@@ -0,0 +1,17 @@
+.section-header{
```
Comment (koops76) [inline chromium/options.css]: Add a space between the `{` and the selector.

File: chromium/options.html [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+<html>
+  <head>
+    <meta charset="utf-8">
+    <link href="options.css" rel="stylesheet" type="text/css"/>
+    <script src="options.js"></script>
+    <script src="translation.js"></script>
+    <script src="send-message.js"></script>
+  </head>
+  <body>
+    <div id='import-confirmed' i18n="options_imported"></div>
```
Comment (koops76) [inline chromium/options.html]: Use `data-i18n` for valid HTML5.

File: chromium/options.html [hunk 17]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <link href="options.css" rel="stylesheet" type="text/css"/>
+    <script src="options.js"></script>
```
Comment (koops76) [inline chromium/options.html]: Move the scripts to the end of `body` tag.

File: chromium/options.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
+document.addEventListener("DOMContentLoaded", () => {
+
+  let json_data;
+  let import_button = document.querySelector("#import");
```
Comment (koops76) [inline chromium/options.js]: `const`?

File: chromium/options.js [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+  function import_json(e) {
+    e.preventDefault();
+
+    let settings = JSON.parse(json_data);
```
Comment (koops76) [inline chromium/options.js]: `const`?

File: chromium/options.js [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    const reader = new FileReader();
+    reader.onload = event => {
+      json_data = event.target.result;
+      import_button.disabled = false;
+    };
+
+    reader.readAsText(file);
+  });
+
+  document.querySelector("form").addEventListener("submit", import_json);
```
Comment (koops76) [inline chromium/options.js]: Maybe find a `form` by `id` instead of simply taking the first one in the document?

File: chromium/options.js [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    e.preventDefault();
+
+    let settings = JSON.parse(json_data);
+    sendMessage("import_settings", settings, resp => {
+      document.querySelector("#import-confirmed").style.display = "block";
+      document.querySelector("form").style.display = "none";
+    });
+  }
+
+  document.querySelector("#import-settings").addEventListener("change", event => {
```
Comment (koops76) [inline chromium/options.js]: `document.getElementById`?

File: chromium/options.js [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+
+  let json_data;
+  let import_button = document.querySelector("#import");
+
+  function import_json(e) {
+    e.preventDefault();
+
+    let settings = JSON.parse(json_data);
+    sendMessage("import_settings", settings, resp => {
+      document.querySelector("#import-confirmed").style.display = "block";
```
Comment (koops76) [inline chromium/options.js]: `document.getElementById`?

File: chromium/options.js [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+      document.querySelector("form").style.display = "none";
```
Comment (koops76) [inline chromium/options.js]: Maybe find a `form` by `id` instead of simply taking the first one in the document?

File: chromium/popup.html [hunk 24]
```diff
@@ -9,6 +9,8 @@
 
     <link href="popup.css" rel="stylesheet" type="text/css"/>
     <script src="popup.js"></script>
+    <script src="translation.js"></script>
```
Comment (koops76) [inline chromium/popup.html]: Move the scripts to the end of `body` tag.

File: chromium/popup.js [hunk 25]
```diff
@@ -197,16 +197,6 @@ document.addEventListener("DOMContentLoaded", function () {
       httpNowhereCheckbox.setAttribute('checked', '');
```
Comment (koops76) [inline chromium/popup.js]: ``httpNowhereCheckbox.checked = true`?

File: chromium/popup.js [hunk 26]
```diff
@@ -197,16 +197,6 @@ document.addEventListener("DOMContentLoaded", function () {
...
     }
   });
-
-  // auto-translate all elements with i18n attributes
-  var elem = document.querySelectorAll("[i18n]");
```
Comment (koops76) [inline chromium/popup.js]: Use `document.querySelectorAll("[data-i18n]")` instead.

File: chromium/popup.js [hunk 27]
```diff
@@ -197,16 +197,6 @@ document.addEventListener("DOMContentLoaded", function () {
...
-  for (let el of elem) {
-    el.innerHTML = chrome.i18n.getMessage(el.getAttribute("i18n"));
```
Comment (koops76) [inline chromium/popup.js]: Use `el.getAttribute("data-i18n")` or `el.dataset.i18n`.

**Comment by Hainish** [comment]
I figured it out....

File: chromium/popup.js [hunk 28]
```diff
@@ -197,16 +197,6 @@ document.addEventListener("DOMContentLoaded", function () {
...
-  }
-
-  // other translations
-  e("aboutTitle").setAttribute("title", chrome.i18n.getMessage("about_title"));
```
Comment (koops76) [inline chromium/popup.js]: `e("aboutTitle").title = chrome.i18n.getMessage("about_title")`

File: chromium/rules.js [hunk 29]
```diff
@@ -185,6 +185,14 @@ function RuleSets(ruleActiveStates) {
 
   // A hash of rule name -> active status (true/false).
   this.ruleActiveStates = ruleActiveStates;
+
+  // Load legacy_custom_rulesets
+  const that = this;
```
Comment (koops76) [inline chromium/rules.js]: ???

File: chromium/rules.js [hunk 30]
```diff
@@ -185,6 +185,14 @@ function RuleSets(ruleActiveStates) {
...
+  storage.get({legacy_custom_rulesets: []}, item => {
+    for(let legacy_custom_ruleset of item.legacy_custom_rulesets){
```
Comment (koops76) [inline chromium/rules.js]: `const`?

File: chromium/translation.js [hunk 31]
```diff
@@ -0,0 +1,11 @@
+document.addEventListener("DOMContentLoaded", function () {
+  // auto-translate all elements with i18n attributes
+  var elem = document.querySelectorAll("[i18n]");
+  for (let el of elem) {
+    el.innerHTML = chrome.i18n.getMessage(el.getAttribute("i18n"));
+  }
+
+  // other translations
+  e("aboutTitle").setAttribute("title", chrome.i18n.getMessage("about_title"));
```
Comment (koops76) [inline chromium/translation.js]: `e("aboutTitle").title = chrome.i18n.getMessage("about_title")`

File: chromium/translation.js [hunk 32]
```diff
@@ -0,0 +1,11 @@
+document.addEventListener("DOMContentLoaded", function () {
+  // auto-translate all elements with i18n attributes
+  var elem = document.querySelectorAll("[i18n]");
+  for (let el of elem) {
+    el.innerHTML = chrome.i18n.getMessage(el.getAttribute("i18n"));
```
Comment (koops76) [inline chromium/translation.js]: `el.innerText = chrome.i18n.getMessage(el.getAttribute("data-i18n"));`

File: chromium/translation.js [hunk 33]
```diff
@@ -0,0 +1,11 @@
+document.addEventListener("DOMContentLoaded", function () {
+  // auto-translate all elements with i18n attributes
+  var elem = document.querySelectorAll("[i18n]");
```
Comment (koops76) [inline chromium/translation.js]: `const elements = document.querySelectorAll("[data-i18n]");`

File: chromium/translation.js [hunk 34]
```diff
@@ -0,0 +1,11 @@
...
+  for (let el of elem) {
```
Comment (koops76) [inline chromium/translation.js]: `const`?

File: src/bootstrap.js [hunk 35]
```diff
@@ -0,0 +1,118 @@
+"use strict";
+
+const CI = Components.interfaces;
+const CC = Components.classes;
+
+function get_prefs(branch_name){
+  let o_prefs = CC["@mozilla.org/preferences-service;1"]
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 36]
```diff
@@ -0,0 +1,118 @@
...
+                      .getService(CI.nsIPrefService);
+  let o_branch = o_prefs.getBranch(branch_name);
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 37]
```diff
@@ -0,0 +1,118 @@
...
+  return o_branch;
+}
+
+function get_custom_rulesets_array(){
+  var loc = "ProfD";  // profile directory
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 38]
```diff
@@ -0,0 +1,118 @@
...
+  var dir =
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 39]
```diff
@@ -0,0 +1,118 @@
...
+    .getService(CI.nsIProperties)
+    .get(loc, CI.nsILocalFile)
+    .clone();
+  dir.append("HTTPSEverywhereUserRules");
+  // Check for existence, if not, create.
+  if (!dir.exists()) {
+    return [];
+  }
+
+  var entries = dir.directoryEntries;
```
Comment (koops76) [inline src/bootstrap.js]: `const`/`let`?

File: src/bootstrap.js [hunk 40]
```diff
@@ -0,0 +1,118 @@
...
+  var files = [];
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 41]
```diff
@@ -0,0 +1,118 @@
...
+  while(entries.hasMoreElements()) {
```
Comment (koops76) [inline src/bootstrap.js]: `while (`

File: src/bootstrap.js [hunk 42]
```diff
@@ -0,0 +1,118 @@
...
+    var entry = entries.getNext();
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 43]
```diff
@@ -0,0 +1,118 @@
...
+    entry.QueryInterface(CI.nsIFile);
+    files.push(entry);
+  }
+
+  let file_data = []
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 44]
```diff
@@ -0,0 +1,118 @@
...
+  for(let file of files){
+    var data = "";
```
Comment (koops76) [inline src/bootstrap.js]: `let`?

File: src/bootstrap.js [hunk 45]
```diff
@@ -0,0 +1,118 @@
...
+    var fstream = CC["@mozilla.org/network/file-input-stream;1"]
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 46]
```diff
@@ -0,0 +1,118 @@
...
+        .createInstance(CI.nsIFileInputStream);
+    var sstream = CC["@mozilla.org/scriptableinputstream;1"]
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 47]
```diff
@@ -0,0 +1,118 @@
...
+        .createInstance(CI.nsIScriptableInputStream);
+    fstream.init(file, -1, 0, 0);
+    sstream.init(fstream);
+
+    var str = sstream.read(4096);
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 48]
```diff
@@ -0,0 +1,118 @@
...
+
+  return file_data;
+}
+
+function startup({webExtension}) {
+  webExtension.startup().then(api => {
+    const {browser} = api;
+    browser.runtime.onMessage.addListener((msg, sender, sendReply) => {
+      if (msg == "import-legacy-data") {
+        let globals = get_prefs("extensions.https_everywhere.");
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 49]
```diff
@@ -0,0 +1,118 @@
...
+        if(!pref_retrieve_bool_with_default(globals, "webextension-migrated", false)){
+          let changed = false;
+
+          let prefs = {
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 50]
```diff
@@ -0,0 +1,118 @@
...
+          };
+
+          if(prefs.http_nowhere_enabled == true ||
+            prefs.global_enabled == false ||
+            prefs.show_counter == true
+          ){
+            changed = true;
+          }
+
+          let rule_toggle = {}
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 51]
```diff
@@ -0,0 +1,118 @@
...
+          let rules = get_prefs("extensions.https_everywhere.rule_toggle.");
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 52]
```diff
@@ -0,0 +1,118 @@
...
+          for(let rule_toggle_key of rules.getChildList("", {})){
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

File: src/bootstrap.js [hunk 53]
```diff
@@ -0,0 +1,118 @@
...
+            rule_toggle[rule_toggle_key]  = rules.getBoolPref(rule_toggle_key);
+            changed = true;
+          }
+
+          let custom_rulesets = get_custom_rulesets_array();
```
Comment (koops76) [inline src/bootstrap.js]: `const`?

**Comment by Hainish** [comment]
@jsha There's a problem with loading the XML for legacy rulesets in the constructor.  The rules themselves aren't stored in `localStorage` and are only loaded on what's currently line 31 of this branch in `background.js`.  So in order to just do `new RuleSets(localStorage)` any time we want to reload rules, we'd have to move that into the constructor as well (along with the lookup for mixed content rules).  I'm uncomfortable with this, because it makes the RuleSets object less agnostic, and we will have to rip this out anyway when we merge the `sign-rulesets` branch, which loads rules and stores them via the network.

I therefore propose that we keep the previous workflow for loading rulesets.

**Comment by jsha** [comment]
That sounds fine for now, although I think the localStorage distinction is a bit of a red herring, since the custom rules get stored in the `chrome.storage` API, which nearly the same thing but a little fancier. I would rephrase it as: The current "load custom rulesets" code expects to be able to call `.addFromXml` on a fully constructed RuleSets object. Calling it from the middle of the constructor winds up being a bit weird. Might be worth refactoring, but not on the blocking path for this PR IMO.

LGTM, modulo @koops76's feedback, and fixing the tests.

**Comment by Hainish** [comment]
@jsha, great, I'll change it back to the way it was and merge master into this branch.  Can you give me a final LGTM before I pull the trigger?

**Comment by jsha** [comment]
Sure, will check back later tonight.

**Comment by Hainish** [comment]
@koops76 thanks for the feedback.  Since we are trying to get this out the door, I'd like to save that for a further refactor.

**Comment by koops76** [comment]
@Hainish No problem.

File: chromium/background.js [hunk 54]
```diff
@@ -660,5 +682,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+          if(ruleset.name == ruleset_name){
+            ruleset.active = ruleset_active;
+            return;
+          }
+        }
+      }
+    })(ruleset_name, settings.rule_toggle[ruleset_name]);
+
+    // This line is necessary to clear cached lookups of rulesets already loaded
+    all_rules.ruleCache = new Map();
```
Comment (jsha) [inline chromium/background.js]: I think you reverted too far. You had an intermediate branch that looked good where you skipped the multiply-nested for loops above, and called `all_rules = new RuleSet(localStorage)`. That will work fine if you just much `load_legacy_custom_rulesets(...)` below `new RuleSet`

Comment (Hainish) [inline chromium/background.js]: @jsha we'd have to copy lines 27-32 as well...

```js
// Allow users to enable `platform="mixedcontent"` rulesets                
var enableMixedRulesets = false;                                           
storage.get({enableMixedRulesets: false}, function(item) {                 
  enableMixedRulesets = item.enableMixedRulesets;                          
  all_rules.addFromXml(loadExtensionFile('rules/default.rulesets', 'xml'));
});                                                                        
```

Comment (Hainish) [inline chromium/background.js]: Otherwise we'll have the problem I mentioned in https://github.com/EFForg/https-everywhere/pull/11792#issuecomment-321417318.

Comment (Hainish) [inline chromium/background.js]: We could do

```js
all_rules = new RuleSet(localStorage);
all_rules.addFromXml(loadExtensionFile('rules/default.rulesets', 'xml'));
load_legacy_custom_rulesets(settings.custom_rulesets);
```

Which is probably fine for now, but seems like we may run into complications when we are working on merging `sign-rulesets`.  Though I can take care of that problem when we come to it, I suppose.

**Comment by Hainish** [comment]
@jsha ready for final review

File: chromium/background.js [hunk 55]
```diff
@@ -660,5 +682,47 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
...
+chrome.runtime.sendMessage("import-legacy-data", import_settings);
+
+/**
+ * Import extension settings (custom rulesets, ruleset toggles, globals) from an object
+ * @param settings the settings object
+ */
+async function import_settings(settings){
+  if(settings.changed){
+    // Load custom rulesets and add to storage
+    await new Promise(resolve => {
```
Comment (koops76) [inline chromium/background.js]: Do we really need to wait until the settings are stored?

Comment (Hainish) [inline chromium/background.js]: @koops76 we do for the options ui.  We only want to display the `Settings imported` string once the settings have actually been imported.

**Comment by koops76** [comment]
@Hainish My opinion is: it's OK, but has repetitive code that must be moved into a function and you probably do not need to wait until the settings are stored, or instead you could use `Promise.all`

File: chromium/background.js [hunk 56]
```diff
@@ -207,8 +212,21 @@ AppliedRulesets.prototype = {
     return null;
   },
 
+  addHostnameToTab: function(tabId, ruleset_name, hostname) {
```
Comment (jsha) [inline chromium/background.js]: Looks like the "addHostnameToTab" work snuck back in by accident.

File: chromium/background.js (hunk 55 above)
Comment (koops76) [inline chromium/background.js]: Then use `Promise.all` to await all four promises.

File: chromium/background.js (hunk 56 above)
Comment (Hainish) [inline chromium/background.js]: @jsha this is just because this was forked from the first iteration of `firefox-webextensions-core` - should I rebase it?

Comment (Hainish) [inline chromium/background.js]: I could rebase from `master`, now that `firefox-webextensions-core` is merged.

Comment (koops76) [inline chromium/background.js]: @Hainish Rebase it I guess.

**Comment by Hainish** [comment]
New PR, rebased from master: https://github.com/EFForg/https-everywhere/pull/11800

**Comment by jsha** [comment]
I'll review that, but it probably would have been easier to merge rather than rebase.

**Comment by Hainish** [comment]
The thing is that this PR was a request to merge in to a branch that was eventually rebased before being merged into master, so the whole history got a bit unwieldy. Let's try doing all merges next time.

On August 9, 2017 8:02:49 PM PDT, Jacob Hoffman-Andrews <notifications@github.com> wrote:
>I'll review that, but it probably would have been easier to merge
>rather than rebase.
>
>-- 
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/EFForg/https-everywhere/pull/11792#issuecomment-321437766

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
